### PR TITLE
Application drafts 3/3

### PR DESCRIPTION
**Pull request by mauditecandela**
This PR compares the applications-edit with the applications-draft branch.
- Allows the user to save drafts before submitting an application. This may help them to prepare the application before they send it out
1. In order to do this, we created a new migration that adds a new field to the applications table called submitted with a boolean value. If the application is submitted means that is not a draft
2. The default value for application.submitted is false. **Note** We need to modify this value in all applications that are currently in the production database as they are all submitted
3. This value is defined by how the user decides to save their application, if they click on "Submit button" or "Save draft"

- We only allow the users to save one application by event. We do this by not displaying the button "Apply" in the frontend but also checking it in the controller. 
- We allow the users to submit an application from a draft by clicking the submit button. We don't validate there email_confirmation or terms_and_conditions because we already did that in the draft. This way we don't need to include all the form fields in the submission.
- We also added a view so the user can see all their applications: Submitted ones and drafts.
- We created a new mailer that warns the applicants about an event deadline coming up (2 days) if they have a draft for that event saved. We need to add this task to Heroku.
- Adds a name for the user so this information can be used in the applications -- making it easier for them to apply for different conferences.
- We created tests for the whole implementation.

Note: We also added a redirection for the users after they sign in. They will be redirected to the previous page they were browsing before signing in.

### CODE REVIEW

File: app/views/events/_index_list_item.erb [hunk 1]
```diff
@@ -12,13 +12,26 @@
...
   </p>
-
   <% if event.application_process == 'application_by_organizer' %>
     <%= button_to "Apply at #{event.name}", event.application_link, method: :get,
         class: "btn btn-save btn-external", disabled: event.closed? %>
-  <% else %>
+  <% elsif current_user == nil || event.applications.find_by(applicant_id: current_user.id) == nil %>
     <%= button_to 'Apply', new_event_application_path(event_id: event.id), method: :get,
         class: "btn btn-save", disabled: event.closed? %>
+  <% elsif event.applications.find_by(applicant_id: current_user.id).submitted %>
```
Comment (mauditecandela) [inline app/views/events/_index_list_item.erb]: ternary operator for defining the button text and use the same code?

File: app/views/events/show.html.erb [hunk 2]
```diff
@@ -12,9 +12,25 @@
   <% end %>
 
   <% if @event.application_process == 'application_by_organizer' %>
-    <%= button_to "Apply at #{@event.name}", @event.application_link, method: :get, class: 'btn btn-save btn-external', disabled: @event.closed? %>
+    <%= button_to "Apply at #{@event.name}", @event.application_link, method: :get,
+        class: "btn btn-save btn-external", disabled: @event.closed? %>
+  <% elsif current_user == nil || @event.applications.find_by(applicant_id: current_user.id) == nil %>
+    <%= button_to 'Apply', new_event_application_path(params[:id]), method: :get,
+        class: "btn btn-save", disabled: @event.closed? %>
+  <% elsif @event.applications.find_by(applicant_id: current_user.id).submitted %>
```
Comment (mauditecandela) [inline app/views/events/show.html.erb]: Also here, ternary operator?

File: config/routes.rb [hunk 3]
```diff
@@ -10,7 +10,7 @@
   end
 
   resources :events do
-    resources :applications, only: [:edit, :update, :show, :new, :create, :destroy]
+    resources :applications, only: [:edit, :update, :show, :new, :create, :destroy, :admin_destroy]
```
Comment (mauditecandela) [inline config/routes.rb]: :admin_destroy is not needed here

File: test/controllers/applications_controller_test.rb [hunk 4]
```diff
@@ -24,18 +24,30 @@ class ApplicationsControllerTest < ActionController::TestCase
...
     end
 
-    it 'shows application to admin users' do
+    it 'shows submitted application to admin users' do
       user = make_user(admin: true)
       sign_in_as(user)
 
       event = make_event
-      application = make_application(event)
+      application = make_application(event, submitted: true)
```
Comment (mauditecandela) [inline test/controllers/applications_controller_test.rb]: @guainumbi is this submitted: true still needed with the new helper method?

File: test/controllers/applications_controller_test.rb [hunk 5]
```diff
@@ -24,18 +24,30 @@ class ApplicationsControllerTest < ActionController::TestCase
...
 
       assert_response :success
     end
 
+    it 'redirects admin users to root if they try to see drafted applications' do
+      user = make_user(admin: true)
+      sign_in_as(user)
+
+      event = make_event
+      application = make_application(event, submitted: false)
```
Comment (mauditecandela) [inline test/controllers/applications_controller_test.rb]: make_draft?

File: app/mailers/applicant_mailer.rb [hunk 6]
```diff
@@ -5,4 +5,9 @@ def application_received(application)
     mail(to: @application.email, subject: "Your application for #{@application.event.name}.")
 	end
 
+	def deadline_reminder(application)
+		@application = application
```
[the remark this answers is not in the document]
Comment (mauditecandela) [reply app/mailers/applicant_mailer.rb]: I think it's an optical illusion! Is this real life or is this just fantasy?

[the remark this answers is not in the document]
Comment (guainumbi) [reply app/mailers/applicant_mailer.rb]: O.O there were just two blank lines bottom and top i took them away.

**Pull request merged**